tensorflow/core/kernels/scatter_functor.h

```diff
@@ -217,6 +258,30 @@ struct ScatterFunctorBase {
     }
     return -1;
   }
+
+  Index operator()(OpKernelContext* c, const Device& d,
+                   typename TTypes<T>::Matrix params,
+                   typename TTypes<T>::ConstMatrix updates,
+                   typename TTypes<Index>::ConstFlat indices) {
+    // indices and params sizes were validated in DoCompute().
```

Comment:
I tried this PR internally, and it's still significantly slower than the serial version. (~15% slower for small cases, up to 16x slower for large cases.) I still don't have time to look at it. So sorry about that! To not block you further, Let's use the parallel code only when `PLATFORM_GOOGLE` is not defined. Could you please wrap the code with `#ifdef PLATFORM_GOOGLE`?
```
#ifdef PLATFORM_GOOGLE
    // The parallel version is significantly slower internally. Only call the
    // serial version for now.
    // TODO(penporn): Avoid locking in parallelization (sort beforehand).
    return SerialExecute(c, d, params, updates, indices);
#else
     // indices and params sizes were validated in DoCompute().
    const Index N = static_cast<Index>(indices.size());
    const Index limit = static_cast<Index>(params.dimension(0));
    ...
#endif  // PLATFORM_GOOGLE
```

Reply:
Hi @penpornk , Are you running the benchmarks in scatter_op_test? I  ran the test, with the PR and without the PR using clang and this is what i see.
With PR
BM_ScatterAddInt32/1               3112     232906       321.4M items/s
BM_ScatterAddInt32/10              5009     100000       1996.4M items/s
BM_ScatterAddInt32/64             14396      45221       4445.6M items/s
BM_ScatterAddInt32/256            94204       7372       2717.5M items/s
BM_ScatterAddInt32/1024          327392       2186       3127.7M items/s
BM_ScatterAddInt32Large/1       7622000        100       131.2M items/s
BM_ScatterAddInt32Large/10      7225810        100       1383.9M items/s
BM_ScatterAddInt32Large/64     10714610        100       5973.2M items/s
BM_ScatterAddInt32Large/256    38700000        100       6615.0M items/s
BM_ScatterAddInt32Large/1024  105354390        100       9719.6M items/s

No PR (serial execute always)

BM_ScatterAddInt32/1               3132     222999       319.2M items/s
BM_ScatterAddInt32/10              5003     141062       1998.7M items/s
BM_ScatterAddInt32/64             13886      50565       4609.1M items/s
BM_ScatterAddInt32/256            96351       7565       2657.0M items/s
BM_ScatterAddInt32/1024          327748       2168       3124.4M items/s
BM_ScatterAddInt32Large/1       9867540        100       101.3M items/s
BM_ScatterAddInt32Large/10     20585780        100       485.8M items/s
BM_ScatterAddInt32Large/64     51982330        100       1231.2M items/s
BM_ScatterAddInt32Large/256   165220150        100       1549.4M items/s
BM_ScatterAddInt32Large/1024  602341830        100       1700.0M items/s

I used this command line. 
bazel run --config=download_clang -c opt //tensorflow/core/kernels:scatter_op_test -- --benchmarks=..
Please let me know if anything else is missed.

Reply:
we were also thinking along the lines of sorting to improve memory locality but did not evaluate that option.

Reply:
Thank you for testing with clang! Yes, I see similar speedups with external TF with clang. The problem with the internal one (i.e., same TF code built with internal toolchain) is with threading. 
Serial scatter: Internal consistently faster than external (clang) by a constant factor.
Parallel scatter: Internal is so much slower than its external counterpart or even the internal serial scatter (up to 16x slower than serial).

I'm not even sure if this is the locking overhead. Might just be threading overhead too. (We link to a different threading library internally.) So I don't think you'd be able to replicate the issue. :( I don't have time to look into this yet (not in the next two weeks). If you'd like to get it in soon then the best way is to just wrap it with `PLATFORM_GOOGLE`. This way normal TensorFlow (external) will benefit from your optimization too.